design-proposals/klip-50-partition-and-offset-in-ksqldb.md

```diff
@@ -0,0 +1,253 @@
+  which will cause ksqlDB to copy the selected pseudocolumn(s) into the output _value_
+* `ROWPARTITION` and `ROWOFFSET` will become reserved column names, and users will not
+  be allowed to use these names either as source column names or as output column aliases.
+  (Existing persistent queries that already use these names will continue to run undisrupted.
+  See ["Compatibility Implications"](#compatibility-implications) for more.)
+
+Consistent with `ROWTIME`, the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns will 
+also be allowed in `WHERE`, `GROUP BY`, and `PARTITION BY` clauses, though use cases
+for the latter two are probably limited. `ROWPARTITION` and `ROWOFFSET` will also be
+allowed in UDFs.
+
+### Pull Queries
+
+Today, users are able to select the `ROWTIME` pseudocolumn as part of pull queries:
+```
+SELECT *, ROWTIME FROM my_table WHERE my_key='foo';
+```
+For multiple reasons, we propose that the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns
+should diverge from this behavior. In other words, users will NOT be allowed to select
+`ROWPARTITION` or `ROWOFFSET` as part of a pull query.
+
+The first reason is that selecting `ROWPARTITION` or `ROWOFFSET` for a table
+is semantically dubious, especially for tables that are the result of other
+computation such as an aggregation over a stream. Should `ROWPARTITION` and `ROWOFFSET`
+in such a case represent the partition and offset of the message in the source stream
+that resulted in the aggregate value, or should they represent the partition and offset
+of the aggregate value in the output table? Note that timestamps don't have this
+ambiguity as the timestamp of the two records mentioned are the same.
+
+The second reason has to do with implications on state store size. ksqlDB is able to 
+support `ROWTIME` in pull queries because the state stores from which pull queries are 
+served includes the timestamp for the latest table row (for each key). In contrast, 
+message partition and offset are not saved in the state stores. As a result, supporting 
+`ROWPARTITION` and `ROWOFFSET` in pull queries would require adding this information
+to state stores. (I think there might be a way to get the partition by other means,
+from the replica queried in order to serve the pull query, but offset information
+definitely wouldn't be available.)
+
+If a user really wants to make partition and/or offset information available for pull
+queries, they can choose to explicitly include the source partition and offset as
+part of the query that materializes their table and query for partition/offset as
+value columns as part of their pull query.
+
+### Other semantics
+
+Similar to `ROWTIME`, the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns will not 
+be included in the list of columns returned as part of a source description.
+
+`ROWPARTITION` and `ROWOFFSET` will also not be added to the output of `PRINT TOPIC` commands.
```

Comment:
partition is already there (https://github.com/confluentinc/ksql/pull/6641) - it might make sense to add offset as well, even if it isn't part of the "print" schema

Reply:
Fascinating, and good to know! I'll move this to be out of scope for this KLIP since it's a bit orthogonal, but I agree it could be nice to add.